src/linker/Linker.Steps/RemoveUnreachableBlocksStep.cs

```diff
@@ -16,72 +16,57 @@ namespace Mono.Linker.Steps
+		struct ProcessingNode
 		{
-			var assemblies = Context.Annotations.GetAssemblies ().ToArray ();
-
-			constExprMethods = new Dictionary<MethodDefinition, Instruction> ();
-
-			do {
-				//
-				// Body rewriting can produce more methods with constant expression
-				//
-				constExprMethodsAdded = false;
-
-				foreach (var assembly in assemblies) {
-					if (Annotations.GetAction (assembly) != AssemblyAction.Link)
-						continue;
-
-					RewriteBodies (assembly.MainModule.Types);
-				}
-			} while (constExprMethodsAdded);
+			public MethodDefinition Method;
+			public int LastAttemptStackVersion;
 		}
 
-		bool TryGetConstantResultInstructionForMethod (MethodDefinition method, out Instruction constantResultInstruction)
-		{
-			if (constExprMethods.TryGetValue (method, out constantResultInstruction))
-				return constantResultInstruction != null;
-
-			constantResultInstruction = GetConstantResultInstructionForMethod (method, instructions: null);
-			constExprMethods.Add (method, constantResultInstruction);
-
-			return constantResultInstruction != null;
-		}
+		// Stack of method nodes which are currently being processed.
+		// Implemented as linked list to allow easy referal to nodes and efficient moving of nodes within the list.
+		// The top of the stack is the first item in the list.
+		LinkedList<ProcessingNode> processingStack;
+
+		// Each time an item is added or removed from the processing stack this value is incremented.
+		// Moving items in the stack doesn't increment.
+		// This is used to track loops - if there are two methods which have dependencies on each other
+		// the processing needs to detect that and mark at least one of them as nonconst (regardless of the method's body)
+		// to break the loop.
+		// This is done by storing the version of the stack on each method node when that method is processed,
+		// if we get around to process the method again and the version of the stack didn't change, then there's a loop
+		// (nothing changed in the stack - order is unimportant, as such no new information has been added and so
+		// we can't resolve the situation with just the info at hand).
+		int processingStackVersion;
+
+		// Stores results of method processing. This state is kept forever to avoid reprocessing of methods.
+		// If method is not in the dictionary it has not yet been processed and it has not been added processing either.
+		// The value in this dictionary can be
+		//   - Reference to the linked list node in the processingStack if the method is on the stack and not yet processed.
+		//   - ProcessedUnchangedSentinel - which means the method has been fully processed and nothing was changed on it - its value is unknown
+		//   - NonConstSentinel - which means the method has been processed and the return value is not a const
+		//   - Instruction instance - method has been processed and it has a constant return value (the value of the instruction)
+		Dictionary<MethodDefinition, object> processedMethods;
```

Comment:
The only value which we could in theory store in a different place is the reference to the queue. All the others are valid "final" results. The instruction means it's a const method and has a value (just like before). The `NonConstSentinel` means it doesn't have a const value - this is new but necessary. Before this change we relied on multiple iterations to react to changes. The cache only stored const values - if a method was not in the cache it could mean either "nonconst" or "don't know/yet". In this change we need to do things per-method, and so we need to store the negative "nonconst" results in the cache as well.
The only optimization here is the `ProcessedUnchangedSentitel` which means "processed, but didn't compute result". This is pretty important optimization. In the hello world app, there's a total off 75K methods, 67K end up in this `ProcessedUnchanged` state - meaning we did process them (good), but nothing needed to their value. So we avoided running the const analyzer on ~80% of the methods. I could remove this, but it feels like a good optimization.

Reply:
What I meant is that the code uses the cache for two different purpose

1. as the final method -> Instructions? cache
2. as temporary processing cache

and I'm just not sure it's worth dropping to object to facilitate that

Reply:
Makes sense - split it into two dictionaries.